### PR TITLE
CSD-72 solo importer

### PR DESCRIPTION
**Pull request by pookmish**
# READY FOR REVIEW

# Summary
- Added solo importer

# Need Review By (Date)
- 4/18

# Urgency
- medium

# Steps to Test
1. `composer require SU-HKKU/cardinal_service_profile:dev-CSD-72-solo-importer`
1. install a new site or run config import (`drush cim`)
1. run the importer `drush @cardinalservice.local mim solo_opportunities`
1. go to `/admin/content` and view an opportunity
1. validate some fields are populated.

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

**Comment by jbickar** [comment]
Sorry for the delay. I have this flagged for review. I am taking the `4/18` literally. I will review it by then.

Of course, that's a Saturday.

**Comment by pookmish** [comment]
poke

**Comment by jbickar** [comment]
_FINE_

**Comment by jbickar** [comment]
```
spr-jbickar-9: ~/Documents/working/ACE/gryphon$ lando drush @cardinalservice.local mim solo_opportunities
  39/163 [======>---------------------]  23%libpng warning: iCCP: extra compressed data
  64/163 [==========>-----------------]  39%libpng warning: iCCP: known incorrect sRGB profile
  84/163 [==============>-------------]  51%libpng warning: iCCP: known incorrect sRGB profile
 155/163 [==========================>-]  95%libpng warning: iCCP: extra compressed data
 157/163 [==========================>-]  96% [error]  ImageMagick error 1: identify-im6.q16: Expected 8192 bytes; found 5437 bytes `/app/docroot/sites/cardinalservice/files/media/image/solo/jenniferpham.png' @ warning/png.c/MagickPNGWarningHandler/1654.
identify-im6.q16: Read Exception `/app/docroot/sites/cardinalservice/files/media/image/solo/jenniferpham.png' @ error/png.c/MagickPNGErrorHandler/1628.
identify-im6.q16: Expected 8192 bytes; found 5437 bytes `/app/docroot/sites/cardinalservice/files/media/image/solo/jenniferpham.png' @ warning/png.c/MagickPNGWarningHandler/1654.
identify-im6.q16: Read Exception `/app/docroot/sites/cardinalservice/files/media/image/solo/jenniferpham.png' @ error/png.c/MagickPNGErrorHandler/1628.
identify-im6.q16: Expected 8192 bytes; found 5437 bytes `/app/docroot/sites/cardinalservice/files/media/image/solo/jenniferpham.png' @ warning/png.c/MagickPNGWarningHandler/1654.
identify-im6.q16: Read Exception `/app/docroot/sites/cardinalservice/files/media/image/solo/jenniferpham.png' @ error/png.c/MagickPNGErrorHandler/1628.
identify-im6.q16: unknown image property "%[profiles]" @ warning/property.c/InterpretImageProperties/3831.
identify-im6.q16: unknown image property "%[EXIF:Orientation]" @ warning/property.c/InterpretImageProperties/3831.
 [command: identify -format 'format:%[magick]|width:%[width]|height:%[height]|colorspace:%[colorspace]|profiles:%[profiles]|exif_orientation:%[EXIF:Orientation]\n' '/app/docroot/sites/cardinalservice/files/media/image/solo/jenniferpham.png'] 
libpng warning: iCCP: known incorrect sRGB profile
 [warning] imagecreatefrompng(): gd-png:  fatal libpng error: Read Error: truncated data ColorMean.php:249
 [warning] imagecreatefrompng(): gd-png error: setjmp returns error condition ColorMean.php:249
 [warning] imagecreatefrompng(): 'public://media/image/solo/jenniferpham.png' is not a valid PNG file ColorMean.php:249
 [notice] Processed 163 items (161 created, 0 updated, 2 failed, 0 ignored) - done with 'solo_opportunities'

In MigrateToolsCommands.php line 870:
                                            
  solo_opportunities Migration - 2 failed.  
                                            
```

**Comment by jbickar** [comment]
Screen Shot 2020-04-17 at 2 24 53 PM

File: config/sync/seckit.settings.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+seckit_xss:
```
Comment (jbickar) [inline config/sync/seckit.settings.yml:1]: This is just ...

.. (wait for it) ...

... _opportunistic_?

File: config/sync/field.storage.node.su_opp_time_year.yml [hunk 2]
```diff
@@ -13,7 +13,7 @@ settings:
   target_type: taxonomy_term
 module: core
 locked: false
-cardinality: 1
+cardinality: -1
```
Comment (jbickar) [inline config/sync/field.storage.node.su_opp_time_year.yml:16]: `-1` is unlimited?

File: tests/codeception/acceptance/OpportunityContentCest.php [hunk 3]
```diff
@@ -37,6 +37,17 @@ public function testOpportunityContentCreation(\AcceptanceTester $I) {
...
 
+  /**
+   * The importer should bring in some content.
+   */
+  public function testImporter(\AcceptanceTester $I) {
+    $I->logInWithRole('administrator');
+    $I->amOnPage('/admin/structure/migrate/manage/opportunities/migrations');
+    $I->canSee('solo_opportunities');
+    $total_items = $I->grabTextFrom('table tbody td:nth-child(4)');
+    $I->assertGreaterOrEquals(1, (int) $total_items);
```
Comment (jbickar) [inline tests/codeception/acceptance/OpportunityContentCest.php:48]: Shouldn't this test run the import first? Or am I missing where it does that?

File: cardinal_service_profile.profile [hunk 4]
```diff
@@ -17,6 +17,19 @@ use Drupal\Core\Site\Settings;
...
+/**
+ * Migration callback to just get the current timestamp.
+ *
+ * @param mixed $arg
+ *   Passed parameter from migration plugin `callback`.
+ *
+ * @return int
+ *   Current timestamp.
+ */
+function cardinal_service_profile_get_time($arg = NULL) {
```
Comment (jbickar) [inline cardinal_service_profile.profile:29]: Where is this used? Why not just use `time()`?

File: config/sync/field.storage.node.su_opp_time_year.yml (hunk 2 above)
Comment (pookmish) [reply config/sync/field.storage.node.su_opp_time_year.yml:16]: yeah.

File: tests/codeception/acceptance/OpportunityContentCest.php (hunk 3 above)
Comment (pookmish) [reply tests/codeception/acceptance/OpportunityContentCest.php:48]: I tried that but the migration takes too long and it runs out of execution time. we'll probably need to work on that later

File: cardinal_service_profile.profile (hunk 4 above)
Comment (pookmish) [reply cardinal_service_profile.profile:29]: thats what you'd think right?! but nope. when you use `time()` it logs messages like "Function accepts 0 arguments but 1 passed" something like that.

File: config/sync/field.storage.node.su_opp_time_year.yml (hunk 2 above)
Comment (jbickar) [reply config/sync/field.storage.node.su_opp_time_year.yml:16]: TY.

File: tests/codeception/acceptance/OpportunityContentCest.php (hunk 3 above)
Comment (jbickar) [reply tests/codeception/acceptance/OpportunityContentCest.php:48]: Hm. Is CircleCI running this test via CodeCeption? 

I see the value of having the test written and included, even if it's not running.

File: cardinal_service_profile.profile (hunk 4 above)
Comment (jbickar) [reply cardinal_service_profile.profile:29]: Where is this being used? PhpStorm isn't giving me anything.

Comment (pookmish) [reply cardinal_service_profile.profile:29]: https://github.com/SU-HKKU/cardinal_service_profile/pull/6/files#diff-dad26c2dac86ef73ece9ba303bf1e20dR119

Comment (jbickar) [reply cardinal_service_profile.profile:29]: Oof.

File: config/sync/migrate_plus.migration.solo_opportunities.yml [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+      type: string
+process:
+  title: title
+  revision_translation_affected:
+    -
+      plugin: default_value
+      default_value: 1
+  revision_timestamp:
+    plugin: callback
+    callable: cardinal_service_profile_get_time
```
Comment (jbickar) [inline config/sync/migrate_plus.migration.solo_opportunities.yml:119]:
```suggestion
    callable: time
```

😂 

_THAT'S A JOKE, SON!_

Comment (pookmish) [reply config/sync/migrate_plus.migration.solo_opportunities.yml:119]: yeah make that change, then run `drush @cardinalservice.local cim -y; drush @cardinalservice.local mim solo_opportunities --update` you'll see the messages

Comment (jbickar) [reply config/sync/migrate_plus.migration.solo_opportunities.yml:119]: I'll believe you on that; don't need to do it myself 😄

File: cardinal_service_profile.profile (hunk 4 above)
Comment (jbickar) [reply cardinal_service_profile.profile:29]: I'll defer to your hard-won experience on this; no need for more dents in my forehead.

Please comment this function extensively based on [your comments here](https://github.com/SU-HKKU/cardinal_service_profile/pull/6#discussion_r410517825). It's clearly a confusing function.

Comment (pookmish) [reply cardinal_service_profile.profile:36]: I added comments on the docblock for this function

File: tests/codeception/acceptance/OpportunityContentCest.php (hunk 3 above)
Comment (pookmish) [reply tests/codeception/acceptance/OpportunityContentCest.php:48]: i added commented out steps for the drush mim execution.

**Pull request merged**